src/rules/indentRule.ts

```diff
@@ -121,6 +156,383 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+     && node.parent.kind === ts.SyntaxKind.ArrowFunction;
+}
+
+function isArrowFunctionWithoutBlock(node: ts.Node): boolean {
+    if (node.parent === undefined) {
+        return false;
+    }
+    return node.kind !== ts.SyntaxKind.Block
+     && node.parent.kind === ts.SyntaxKind.ArrowFunction;
+}
+
+function inStatementParen(node: ts.Node): boolean {
+    if (node.parent === undefined) {
+        return false;
+    }
+    const parentInStatement = [
+        ts.SyntaxKind.IfStatement,
+        ts.SyntaxKind.WhileStatement,
+        ts.SyntaxKind.ForStatement,
+        ts.SyntaxKind.ForOfStatement,
+        ts.SyntaxKind.DoStatement,
+        ts.SyntaxKind.WithStatement,
+    ].indexOf(node.parent.kind) >= 0;
+    return node.kind !== ts.SyntaxKind.Block && parentInStatement;
+}
+
+function isInParen(node: ts.Node, root: ts.SourceFile): boolean {
+    const preChild: ts.Node | undefined = getPreChild(node, root);
+    if (preChild === undefined) {
+        return false;
+    }
+    return preChild.kind === ts.SyntaxKind.OpenParenToken;
+}
+
+function isSameLineWithParent(node: ts.Node, root: ts.SourceFile): boolean {
+    const preChild: ts.Node | undefined = getPreChild(node, root);
+    if (preChild === undefined) {
+        return false;
+    }
+    if (preChild.getText().trim() !== "(") {
+        return false;
+    }
+    return isSameLine(root, node.getStart(), preChild.getEnd());
+}
+
+function getPreChild(node: ts.Node, root: ts.SourceFile): ts.Node | undefined {
+    let preChild: ts.Node | undefined;
+    const parent = node.parent;
+    if (parent === undefined) {
+        return undefined;
+    }
+    const children = parent.getChildren(root);
+    if (children.length === 0) {
+        return undefined;
+    }
+    children.forEach((child) => {
+        if (child.getEnd() <= node.getFullStart() && child.getText().length > 0) {
+            preChild = child;
+        }
+    });
+    return preChild;
+}
+
+function isCloseParen(node: ts.Node, line: string): boolean {
+    if (node.kind !== ts.SyntaxKind.Block) {
+        return false;
+    }
+    return /^\s*\)/.test(line);
+}
+
+function isCloseElement(node: ts.Node, line: string): boolean {
+    if (node.parent === undefined) {
+        return false;
+    }
+    return node.parent.kind === ts.SyntaxKind.JsxElement
+        && /^\s*<\//.test(line);
+}
+
+function leadingWithBinaryOperator(node: ts.Node): boolean {
+    return [
+        ts.SyntaxKind.PlusToken,
+        ts.SyntaxKind.MinusToken,
+        ts.SyntaxKind.AsteriskToken,
+        ts.SyntaxKind.SlashToken,
+        ts.SyntaxKind.AmpersandAmpersandToken,
+        ts.SyntaxKind.BarBarToken,
+    ].indexOf(node.kind) > -1;
+}
+
+/**
+ *  To check a class declaration broken to two lines, like this:
+ *  class
+ *  Foo {
+ *      // xxx
+ *  }
+ *  In this case, the `Foo` is regarded as should have the same indent depth with `class`.
+ *  Should it have one more depth? I'm not sure.
```

Comment:
use proper JSDoc formatting -- there is an extra whitespace after each `*` here

Reply:
done